### PR TITLE
feat: `single_letter_names`

### PR DESCRIPTION
**Pull request by KSXGitHub**
Implements the `single_letter_names` umbrella rule as four independently-toggleable sub-lints:

- `perfectionist::single_letter_generic` — generic type parameters whose name is one ASCII letter, exempted only when owned by a short trait impl block (configurable line threshold, default 20). The exemption is scoped to generics on the impl itself, not generics on inner associated items.
- `perfectionist::single_letter_let_binding` — `let` bindings whose name is one ASCII letter, outside `#[cfg(test)]`. Configurable allowlist (default `["n"]`).
- `perfectionist::single_letter_function_param` — function / method parameters with one-letter names. Configurable allowlist (default `["n", "f", "i", "j", "k"]`).
- `perfectionist::single_letter_closure_param` — closure parameters with one-letter names, except when the body is a single expression and either the closure is the immediate argument of a comparison / fold method or the body is a trivial wrapper around the parameter (field access, method call, one-arg call, or `&param`, with surrounding `*`/`&` operators peeled).

Configuration lives once at `["perfectionist::single_letter_names"]` and is shared across the family, exposing the allowlists, the short-impl line threshold, and the comparison-method allowlist.

Side effect: `gen-docs` now emits one rule per `declare_tool_lint!` invocation rather than only the first one in each module, so sub-lint families surface every entry on the docs page.

### CODE REVIEW

**Comment by KSXGitHub** [comment]
## Review (round 1)

**Overview.** Implements the `single_letter_names` umbrella rule as four sub-lints sharing one configuration block; `gen-docs` now emits one rule per `declare_tool_lint!` invocation so families surface every sub-lint.

**Correctness.** HIR hooks pick the right entry points. `enclosing_short_trait_impl` correctly stops at the first `Item` ancestor (so a fn generic inside an impl is not granted the short-impl exemption). `is_param_ref`'s deref-peel fix is necessary — `(*s).to_owned()` is the shape used by the rule's own `Config::default`, so the pre-fix code failed self-lint.

**Issues I'm acting on.**

1. **Dead code: collected `HirId` never used.** `check_closure_params` builds `Vec<(Ident, HirId)>` and the loop destructures `(ident, _)`. The `binding_hir_id` lookup and the field are dead weight; the local variable also shadows the function name `binding_hir_id`. Cleaning up.
2. **Missing regression test for the deref-peel fix.** `(*s).to_owned()` only fails self-lint, not UI. Adding a UI case so a future refactor can't re-break it silently.
3. **User-facing doc omits the `&param` shape.** `SINGLE_LETTER_CLOSURE_PARAM`'s rustdoc lists three trivial-wrapper shapes (field access, method call, one-arg call); the implementation accepts a fourth via the `AddrOf` branch. Adding the `&param` case to the docstring so an `#[allow]` reader isn't surprised.

**Issues I'm flagging but not acting on.**

- **Trait-impl method param exemption** from the planning file ("skip parameters of trait impl methods whose name comes from the trait signature") is not implemented. The default allowlist (`n`/`f`/`i`/`j`/`k`) absorbs the most common case (`fn fmt(&self, f: ...)`); a first cut without this exemption is defensible but worth flagging.
- **UI coverage gaps**: no boundary test for `short_impl_max_lines`, no test for `i`/`j`/`k` on fn params, no test that a comparison closure with a block body is still flagged. Not blocking.

**Style and surface.** Naming / layout / `CONFIG_KEY` convention / rustdoc heading ("Why restrict this?") all match the existing rules. `gen-docs`'s shared-config-per-family handling is sound; the `Clone` derives are necessary for that.

**Security / performance.** No external input, no unsafe; per-crate HIR walk with small `BTreeSet` lookups, cost is negligible.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by KSXGitHub** [comment]
## Review (round 2)

**Bug found, acting on it.** `enclosing_short_trait_impl` walks `hir_parent_iter` and `continue`s past every node that is not `hir::Node::Item`. That correctly skips `Block`/`Expr` ancestors but also skips `ImplItem`/`TraitItem`, which means a generic on a method inside a short trait impl inherits the impl's exemption:

```rust
impl<T> Foo for Bar {     // short impl
    fn baz<U>(x: U) {}    // `U` should be flagged, but the walk lands on the
                          // impl and grants the short-trait exemption
}
```

The exemption is supposed to apply only to generics owned by the impl block itself. Fixing by stopping the walk at the first `Item`/`ImplItem`/`TraitItem`/`ForeignItem` and only granting the exemption when that ancestor is the impl. Adding a UI fixture for the method-on-short-impl case.

**Minor doc inconsistency, acting on it.** The top-level doc on `is_trivial_wrapper` still lists the four match arms without mentioning the `*`/`&` peel that the inner `is_param_ref` performs. The user-facing rustdoc on `SINGLE_LETTER_CLOSURE_PARAM` is correct after round 1; this is an internal-doc cleanup.

**No regressions from round 1.** Dead `HirId` cleanup landed cleanly, deref-peel UI fixture wired in, `&param` shape now in the user-facing docstring.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by KSXGitHub** [comment]
## Review (round 3) — clean

Re-walked the diff end-to-end after the round-2 fix. No further actionable issues.

**Checked again:**
- `enclosing_short_trait_impl` now stops at the first `Item` / `ImplItem` / `TraitItem` / `ForeignItem`, so the short-impl exemption applies only when the generic is owned by the impl block itself. UI fixture exercises both the trait-method and the impl-method shape.
- `is_trivial_wrapper`'s outer and inner docstrings are consistent: outer lists the four shapes plus the `*`/`&` peel note, inner explains the peel mechanism.
- `check_local` ordering (`is_single_ascii_letter` → allowlist → `is_in_test`) is cheap-first; the allowlist check before `is_in_test` is correct because allowlisted names are exempt regardless of `cfg(test)` context.
- `check_function_params` correctly skips the implicit-self param using `decl.implicit_self`, leaving the rule to apply to remaining positional parameters.
- `single_expression_body` strips a labelled block's label; a labelled-block closure body `|x| 'a: { x.foo() }` is treated as single-expression and recursed into, which matches the spec.

**Out-of-scope flags (still not acting):**
- Trait-impl method param exemption (planning file lists it; current behaviour is to flag every fn param uniformly, with the conventional names absorbed by the default allowlist).
- UI coverage on `short_impl_max_lines` boundary, `i`/`j`/`k` fn-param allowlist, comparison closure with block body — coverage gaps but not bugs.

Stopping the review loop here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/rules/single_letter_names.rs [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+/// Default allowlist of method names whose closure argument may
+/// use single-letter parameters when the body is a single
+/// expression. Both source documents agree on this list.
+const DEFAULT_COMPARISON_METHODS: &[&str] = &[
+    "sort_by",
+    "sort_unstable_by",
+    "sort_by_key",
+    "sort_unstable_by_key",
+    "cmp",
+    "partial_cmp",
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply src/rules/single_letter_names.rs]: Good catch. The list came verbatim from the planning file, which conflated `Ord::cmp` / `PartialOrd::partial_cmp` (take `&Self`) with the closure-accepting `Iterator::cmp_by` / `partial_cmp_by`. Replaced the bare-trait entries with `cmp_by` / `partial_cmp_by` / `eq_by` and refreshed the user-facing lint docstring to match. Pushed in fc38909.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by KSXGitHub** [comment]
## Review (round 4) — clean after Copilot fix

Re-walked the diff after fc38909 (the Copilot fix on `DEFAULT_COMPARISON_METHODS`). No new actionable findings.

**Verified:**
- `DEFAULT_COMPARISON_METHODS` now lists only closure-accepting methods; user-facing docstring matches; inline comment explains why bare `cmp`/`partial_cmp` aren't included.
- `enclosing_short_trait_impl` scopes the impl exemption correctly.
- `is_trivial_wrapper`'s `*`/`&` peel is in place; outer-doc and inner-doc agree.
- `check_local` short-circuits cheap-first.
- UI exercises every sub-lint plus the short-impl-method, deref-peel, and `cfg(test)` exemption corner cases.

**Out-of-scope (unchanged from rounds 2/3):**
- Trait-impl method param exemption (planning-file item, not implemented).
- UI gaps for `i`/`j`/`k` fn-param allowlist and block-body comparison closures.

Stopping the review loop — four convergent rounds, no further findings.

---
_Generated by [Claude Code](https://claude.ai/code)_
